Fix Linux GL Renderer color primaries detection

felixbuenemann opened the pull request:
## Description
The m_iFlags are passed to ColorManager for detecting the color primaries of the video, so it needs to contain the color primary flags for the Colour Management -> Primaries -> Automatic to work as expected.

This is a follow-up to #17109 and fixes desaturated colours for BT.2020 videos that were previously handled at BT.709 because that is the fallback value, when no color primaries flag is set.

## Motivation and Context
After #17109 color management can finally be used with the Linux GL Renderer without crashing. When testing this with 4K HDR videos with BT.2020 primaries I noticed that the colours were completely desaturated. After manually setting Colour Management -> Primaries -> UHDTV (BT.2020), the colors were accurate.

This fix is essential for playing back BT.2020 videos on non-BT.2020 capable displays (eg. sRGB or BT.709).

Since the bug also affects Leia it should be backported, similar to the changes from #17109.

## How Has This Been Tested?
The fix was tested against master@f4f9d88a35 with both software decoding and VAAPI.

The bug was observed without the fix on both Leia stable and nightly builds on Ubuntu 19.10 with Gemini Lake (Intel UHD 605) graphics and both i965 and iHD VAAPI drivers as well as software decoding.

The following settings were used:

* Color management: Enabled
* Colour management mode: ICC Profile
* ICC display profile: (path to monitors ICM profile extracted from windows driver, a standard sRGB.icc should work as well for testing)
* Whitepont: D65
* Gamma mode: BT.1886
* Lookup table size: 4 or 5 (doesn't matter)

The following videos were tested:

* 2160p HDR BT.2020 HEVC MKV
* 1080p SDR BT.709 AVC MKV

For some movies both 4K and Full-HD versions were tested to compare the color timing and the BT.2020 color levels after tone mapping and color management on an Asus PA328Q (100% sRGB IPS) were compared to a Sony Z9D HDR TV as a reference.

Without the fix all BT.2020 content looked completely desaturated with either colour management disabled or set to Primaries -> Automatic, but looked fine with Primaries -> BT.2020. Conversely BT.709 videos looked ok with Automatic, but oversaturated with BT.2020 hinting at the broken auto-detection.

## Screenshots (if appropriate):

Example 1080p SDR BT.709 w/o CM:
![lk_sdr_bt709_no_cm](https://user-images.githubusercontent.com/909587/72572037-9baef700-38c1-11ea-8917-3e9cd73b8dc0.jpg)

Same with CM enabled and automatic primaries:
![lk_sdr_bt709_cm_auto_primaries](https://user-images.githubusercontent.com/909587/72572036-9baef700-38c1-11ea-9280-61f5fb1205e6.jpg)

Example 2160p HDR BT.2020 Reinhard Tone Mapping to SDR w/o CM:
![lk_hdr_bt2020_no_cm](https://user-images.githubusercontent.com/909587/72572035-9baef700-38c1-11ea-92df-a859474dfa13.jpg)

Same with CM enabled and automatic primaries (after fix):
![lk_hdr_bt2020_cm_auto_primaries](https://user-images.githubusercontent.com/909587/72572033-9b166080-38c1-11ea-9656-f394ab5b68f0.jpg)

Same with CM enabled and BT.709 primaries (simulates "automatic" w/o fix):
![lk_hdr_bt2020_cm_bt709_primaries](https://user-images.githubusercontent.com/909587/72572034-9baef700-38c1-11ea-8e90-29a7b160951f.jpg)

Please view screenshots in fullscreen for best results.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed (not tested)

FernetMenta commented:
Primaries a stored in m_srcPrimaries, not in flags. Instead of turning back LinuxRenderer, color manager should be fixed to adapt to the changes in renderer.

FernetMenta commented:
More verbose: color primaries belong to a picture and can change with every picture. flags are only set at configure stage and do not observe changes. m_srcPrimaries are updated.

felixbuenemann commented:
@FernetMenta I just replicated what these renderers are currently doing:

https://github.com/xbmc/xbmc/blob/dee52eafc89569449b1be7eae200c24eb76b1655/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.cpp#L75
https://github.com/xbmc/xbmc/blob/244618bf3a127d879370776a7a0bd4fd7f0bcd40/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererMediaCodecSurface.cpp#L59
https://github.com/xbmc/xbmc/blob/770c691bfeea88d561b39d2e38348ac434db3adc/xbmc/cores/VideoPlayer/VideoRenderers/WinRenderer.cpp#L164
https://github.com/xbmc/xbmc/blob/770c691bfeea88d561b39d2e38348ac434db3adc/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/MMALRenderer.cpp#L934

The only place I found where it seems to be updated for a buffer is here:

https://github.com/xbmc/xbmc/blob/770c691bfeea88d561b39d2e38348ac434db3adc/xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererBase.cpp#L409

Do you propose that all these renderers need to be fixed?

Also can you elaborate or have an example video where the color primaries change on the fly?

Reading the code of the Linux GL renderer it looks like the primaries *could* changhe on the fly, but are there real world cases where that happens?

I'm a bit concerned about the cost of possibly checking if the color manager config has changes for each frame instead of once.

FernetMenta commented:
>  I just replicated what these renderers are currently doing

Bad idea. The authors of those renderers just copied code from and ancient version of LinuxRenderer. MMAL or MediaCodecSurface don't do any color conversions, hence color primaries are useless for them.

> Do you propose that all these renderers need to be fixed?

I don't care. It's just cosmetics, see above.

> Reading the code of the Linux GL renderer it looks like the primaries could changhe on the fly, but are there real world cases where that happens?

It's a valid scenario. You don't even need a change in a single video, just switch a tv channel for example. If dimensions don't change, there is no need to reconfigure the entire renderer.

> I'm a bit concerned about the cost of possibly checking if the color manager config has changes for each frame instead of once.

No need to do so. A change in color primaries trigger ReloadShaders and UpdateVideoFilters. Just feed m_srcPrimaries into colorManager and check for change in CColorManager::CheckConfiguration. From a colorManager perspective primaries just moved from flags to a different parameter.

felixbuenemann commented:
@FernetMenta I've pushed a new version with the requested changes.

The ColorManager now gets passed `AVColorPrimaries` instead of flags.

I've also updated the Windows renderer accordingly, but couldn't test those changes.

So it would be great if someone could test this on Windows.

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:
```diff
@@ -85,21 +84,18 @@ bool CColorManager::IsValid() const
-  if (flags & CONF_FLAGS_COLPRI_240M)
-    return CMS_PRIMARIES_240M;
-  return CMS_PRIMARIES_BT709; // default to bt.709
+  switch(primaries)
```
> Control statement keywords have to be separated from opening parentheses by one space.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:
```diff
@@ -85,21 +84,18 @@ bool CColorManager::IsValid() const
-  return CMS_PRIMARIES_BT709; // default to bt.709
+  switch(primaries)
+  {
+    case AVCOL_PRI_BT709: return CMS_PRIMARIES_BT709;
```
> Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:
```diff
@@ -85,21 +84,18 @@ bool CColorManager::IsValid() const
+  switch(primaries)
+  {
+    case AVCOL_PRI_BT709: return CMS_PRIMARIES_BT709;
+    case AVCOL_PRI_BT470M: return CMS_PRIMARIES_BT470M;
```
> Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:
```diff
@@ -85,21 +84,18 @@ bool CColorManager::IsValid() const
+  {
+    case AVCOL_PRI_BT709: return CMS_PRIMARIES_BT709;
+    case AVCOL_PRI_BT470M: return CMS_PRIMARIES_BT470M;
+    case AVCOL_PRI_BT470BG: return CMS_PRIMARIES_BT470BG;
```
> Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:
```diff
@@ -85,21 +84,18 @@ bool CColorManager::IsValid() const
+    case AVCOL_PRI_BT709: return CMS_PRIMARIES_BT709;
+    case AVCOL_PRI_BT470M: return CMS_PRIMARIES_BT470M;
+    case AVCOL_PRI_BT470BG: return CMS_PRIMARIES_BT470BG;
+    case AVCOL_PRI_SMPTE170M: return CMS_PRIMARIES_170M;
```
> Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:
```diff
@@ -85,21 +84,18 @@ bool CColorManager::IsValid() const
+    case AVCOL_PRI_BT470M: return CMS_PRIMARIES_BT470M;
+    case AVCOL_PRI_BT470BG: return CMS_PRIMARIES_BT470BG;
+    case AVCOL_PRI_SMPTE170M: return CMS_PRIMARIES_170M;
+    case AVCOL_PRI_SMPTE240M: return CMS_PRIMARIES_240M;
```
> Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:
```diff
@@ -85,21 +84,18 @@ bool CColorManager::IsValid() const
+    case AVCOL_PRI_BT470BG: return CMS_PRIMARIES_BT470BG;
+    case AVCOL_PRI_SMPTE170M: return CMS_PRIMARIES_170M;
+    case AVCOL_PRI_SMPTE240M: return CMS_PRIMARIES_240M;
+    case AVCOL_PRI_BT2020: return CMS_PRIMARIES_BT2020;
```
> Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:
```diff
@@ -85,21 +84,18 @@ bool CColorManager::IsValid() const
+    case AVCOL_PRI_SMPTE170M: return CMS_PRIMARIES_170M;
+    case AVCOL_PRI_SMPTE240M: return CMS_PRIMARIES_240M;
+    case AVCOL_PRI_BT2020: return CMS_PRIMARIES_BT2020;
+    default: return CMS_PRIMARIES_BT709; // default to bt.709
```
> Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:
```diff
@@ -152,10 +148,10 @@ bool CColorManager::Get3dLutSize(CMS_DATA_FORMAT format, int *clutSize, int *dat
 }
 
-bool CColorManager::GetVideo3dLut(int videoFlags, int *cmsToken, CMS_DATA_FORMAT format, int clutSize, uint16_t *clutData)
+bool CColorManager::GetVideo3dLut(AVColorPrimaries srcPrimaries, int *cmsToken, CMS_DATA_FORMAT format, int clutSize, uint16_t *clutData)
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.cpp:
```diff
@@ -260,7 +256,7 @@ bool CColorManager::CheckConfiguration(int cmsToken, int flags)
     {
       CMS_PRIMARIES primaries = static_cast<CMS_PRIMARIES>(settings->GetInt("videoscreen.cmsprimaries"));
-      if (primaries == CMS_PRIMARIES_AUTO) primaries = videoFlagsToPrimaries(flags);
+      if (primaries == CMS_PRIMARIES_AUTO) primaries = avColorToCmsPrimaries(srcPrimaries);
```
> Do not put multiple statements on a single line. Always use a new line for a new statement.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#41-multiple-statements

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/ColorManager.h:
```diff
@@ -76,22 +81,22 @@ class CColorManager
    \return true on success, false otherwise
    */
-  bool GetVideo3dLut(int primaries, int *cmsToken, CMS_DATA_FORMAT format, int clutSize, uint16_t *clutData);
+  bool GetVideo3dLut(AVColorPrimaries srcPrimaries, int *cmsToken, CMS_DATA_FORMAT format, int clutSize, uint16_t *clutData);
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp:
```diff
@@ -2670,7 +2670,7 @@ bool CLinuxRendererGL::LoadCLUT()
 
   // load 3DLUT
-  if ( !m_ColorManager->GetVideo3dLut(m_iFlags, &m_cmsToken, CMS_DATA_FMT_RGB, m_CLUTsize, m_CLUT) )
+  if ( !m_ColorManager->GetVideo3dLut(m_srcPrimaries, &m_cmsToken, CMS_DATA_FMT_RGB, m_CLUTsize, m_CLUT) )
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererBase.cpp:
```diff
@@ -314,19 +314,19 @@ bool CRendererBase::CreateIntermediateTarget(unsigned width, unsigned height, bo
 
     const auto lutData = static_cast<uint16_t*>(KODI::MEMORY::AlignedMalloc(dataSize, 16));
-    bool success = m_colorManager->GetVideo3dLut(flags, &m_cmsToken, CMS_DATA_FMT_RGBA, lutSize, lutData);
+    bool success = m_colorManager->GetVideo3dLut(srcPrimaries, &m_cmsToken, CMS_DATA_FMT_RGBA, lutSize, lutData);
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi reviewed (commented):
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

felixbuenemann commented:
@Rechi I've fixed the clang-format violations in newly added code, but kept the existing code as is to avoid obfuscating the changes.

FernetMenta commented:
looks good to me

felixbuenemann commented:
This PR has been stale for a while – is there anything left to do for me here?

phunkyfish commented:
Sorry @felixbuenemann, let me find someone who can test the Windows side. @da-anda can you help?

da-anda commented:
@phunkyfish I can probably test (once I got some HDR test samples since I don't have any local HDR videos), but it might take a while since I'm super stressed out atm

phunkyfish commented:
No rush, thanks @da-anda.

Hope everything is ok.

da-anda commented:
@phunkyfish was about to try this PR, but as it seems that I need to create a 3DLUT file for this (no idea how) I'm not so sure I really want to spend hours trying to figure out what to do and how to test for basically nothing, since I don't use really use Windows for KODI (apart from random testing).

felixbuenemann commented:
@da-anda You don't need a 3d lut, you can use an icc profile from your monitor or a default sRGB.icc. This only makes a difference when colour management is active.

felixbuenemann commented:
Rebased on master at 7091757383c384f7cde853f670346a5ce05f21d5.

da-anda commented:
so I gave this PR a try today. With DXVA2 HW decoding, the colors were correct, regardless if I turned colour management on or off (with random selected ICC profile since i have no idea what the correct one for my notebook screen would be, since this model comes with 4 different screens and the driver comes with like 18 different icm files). I changed all settings from whitepoint to gamma and there was no visual change (while the video kept playing in the background).
Once I have turned off DXVA2 HW acceleration, the sample video was washed out and no colour management setting changed anything - it always looked the same (again, while video kept playing in the background).

Had the same behavior with current Windows nightly. It at least didn't look any different to me. Software decoding was washed out and with DXVA2 colours looked fine.

Do I have to restart KODI for things to take effect, or restart playback? I at least couldn't see a difference, regardless which setting I used.

fritsch commented:
Hi, sorry - that we totally forgot about this PR. We are really sorry. The code is fine. Could you please rebase a final time and check the - really minor - code formating comments (e.g. one statement per line and max char number per line in the hunks, you touch?).

That would be totally great. If you think: Hey, team kodi stop going on my nerves, also fine - just state and I will take it over :-)

Thanks again for helping us out and improving kodi.

felixbuenemann commented:
Sure, I can rebase this later.

Can't say I'm a fan of those automated style checks, most of them decrease readability instead of improving it and they also make it harder to see what actually changed in the diff…

felixbuenemann committed:
Fix Linux GL Renderer color primaries detection

This makes sure the videos source primaries are passed to the color
manager. Also adapted the ColorManager interface to accept AVColorPrimaries
instead of CONF_FLAGS_COLPRI which required updates to the Windows
renderer base.

felixbuenemann commented:
@fritsch Rebased on master at b7e053aebc and fixed clang-format warnings.

fritsch commented:
Thanks very much for investing your time! I will push right away after jenkins is green.

felixbuenemann commented:
Btw. why was the "Backport: Needed" label removed?

Since #17109 was backported to Leia, this should backport fine.

fritsch commented:
The reason is simple: Our Release Manager and the team have decided to not do another Leia release to focus on v19 and release the first alpha very soon.

If this changes, this PR will obviously considered.
